profiling/src/alloc/arena.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright 2023-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+use super::AllocError;
+use crate::alloc::r#virtual::{Mapping, OsVirtualAllocator, VirtualAllocator};
+use core::ptr::NonNull;
+use std::alloc::Layout;
+use std::cell::Cell;
+use std::io;
+
+#[derive(Debug)]
+pub struct ArenaAllocator<A: VirtualAllocator = OsVirtualAllocator> {
+    pub(crate) mapping: Option<Mapping<A>>,
+    first_free_offset: Cell<usize>,
+}
+
+impl<A: VirtualAllocator> Default for ArenaAllocator<A> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl ArenaAllocator<OsVirtualAllocator> {
+    pub fn with_capacity(capacity: usize) -> io::Result<ArenaAllocator<OsVirtualAllocator>> {
+        Self::with_capacity_in(capacity, OsVirtualAllocator {})
+    }
+}
+
+impl<A: VirtualAllocator> ArenaAllocator<A> {
+    /// Creates a new arena allocator which has a capacity of zero. It will
+    /// not request a virtual mapping from the OS.
+    pub const fn new() -> Self {
+        Self {
+            mapping: None,
+            first_free_offset: Cell::new(0),
+        }
+    }
+
+    unsafe fn from_mapping(mapping: Mapping<A>) -> ArenaAllocator<A> {
+        let first_free_offset = Cell::new(0);
+        Self {
+            mapping: Some(mapping),
+            first_free_offset,
+        }
+    }
+
+    /// Creates an arena allocator whose underlying buffer holds at least
+    /// `capacity` bytes. It will round up to a page size, except for capacity
+    /// of zero.
+    pub fn with_capacity_in(capacity: usize, alloc: A) -> io::Result<Self> {
+        if capacity == 0 {
+            return Ok(Self::new());
+        }
+
+        let region = Mapping::new_in(capacity, alloc)?;
+
+        // SAFETY: we haven't done any unsafe things with the region like give
+        // out pointers to its interior bytes.
+        Ok(unsafe { Self::from_mapping(region) })
+    }
+
+    pub fn remaining_capacity(&self) -> usize {
+        match &self.mapping {
+            None => 0,
+            Some(mapping) => mapping.allocation_size() - self.first_free_offset.get(),
+        }
+    }
+
+    /// Allocates the given layout. It will be zero-initialized. Allows for
+    /// zero-sized allocations, which are not guaranteed to have unique
+    /// addresses.
+    pub fn allocate_zeroed(&self, layout: Layout) -> Result<NonNull<[u8]>, AllocError> {
+        let layout = layout.pad_to_align();
+        let size = layout.size();
+        if size == 0 {
+            return Ok(NonNull::from(&mut []));
```

Comment:
Because it's a wide pointer `NonNull<[u8]>`, so we also need size info.

```
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
  --> profiling/src/alloc/arena.rs:70:23
   |
70 |             return Ok(NonNull::dangling());
   |                       ^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `[u8]`
note: required by a bound in `NonNull::<T>::dangling`
  --> /Users/levi.morrison/.rustup/toolchains/1.69.0-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/ptr/non_null.rs:67:6
   |
67 | impl<T: Sized> NonNull<T> {
   |      ^ required by this bound in `NonNull::<T>::dangling`
```